### ISSUE TITLE
T _fcntl$NOCANCEL is most expensive operation; what is this?

### ISSUE THREAD

**Issue by DullReferenceException** [nodejs/help#2190]
* **Node.js Version**: v10.15.3
* **OS**: macOS
* **Scope**: runtime

While profiling my app (an express-based web API), I see that a good 31% of all ticks are taken up by this mysterious player:

```
   ticks  total  nonlib   name
   6772   32.8%   33.1%  T _fcntl$NOCANCEL
```

In the "Bottom up (heavy) profile" section of the profiler output, this operation stands alone with no parent/child breakdown.

What is this mysterious `T _fcntl$NOCANCEL`, and what does it imply is going on in my app?

**Comment by addaleax** [comment]
It’s a POSIX syscall: https://linux.die.net/man/2/fcntl

It means that your app is doing a lot of filesystem meta-operations, but without knowing a bit more it’s hard to tell what exactly those calls do.

On Linux I’d start with looking at the output `strace -f -e fcntl node ...` as that gives a description of what the calls actually do, on macOS there’s dtruss which should also allow you to do that but that I don’t personally know how to use.

**Comment by DullReferenceException** [comment]
Thanks for your tips. I tried playing around with dtruss, but I wasn't able to interpret exactly what things meant. Next I tried removing parts of my code piece by piece to see if I could get `T _fcntl$NOCANCEL` to disappear. Eventually I got it down to just these lines of code:

```js
const http = require('http');
const server = http.createServer((req, res) => {
  res.end('[]');
});
server.listen(8080);
```

When hitting this with tons of requests, I continue seeing this in the profiling data:

```
 [C++]:
   ticks  total  nonlib   name
   1317   25.7%   26.8%  T _fcntl$NOCANCEL
```

...so this must be some core mechanism that takes up all these ticks. The minimal code I put together isn't doing any FS operations to speak of, so is this something socket-related? Event loop related?
